In [1]:
from __future__ import print_function

import numpy as np
import pandas as pd

/Applications/anaconda/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


In [2]:
df = pd.read_csv('/Users/aleksandra/ds/metis/project_luther/project_luther/rotten_tomatoes2.csv')

In [3]:
df.head(2)

url  score_c rating_c  \
0  https://www.rottentomatoes.com/m/resident_evil...       35   4.4/10   
1   https://www.rottentomatoes.com/m/passengers_2016       31   4.9/10   

   n_reviews_c  fresh_c  rotten_c score_u rating_u n_reviews_u  \
0           85       30        55     51%   5.9/10      58,527   
1          229       70       159     63%   4.7/10      52,600   

             director    box_office runtime                 in_theatre  \
0  Paul W.S. Anderson   $26,844,692   P106M  2017-01-26T16:00:00-08:00   
1       Morten Tyldum  $100,014,092   P116M  2016-12-20T16:00:00-08:00   

                     on_disc                                        audience  \
0  2017-05-15T17:00:00-07:00        R (for sequences of violence throughout)   
1  2017-03-13T17:00:00-07:00  PG-13 (for sexuality, nudity and action/peril)   

              studio  
0      Sony Pictures  
1  Columbia Pictures

In [4]:
len(df)

1484

In [5]:
df.dtypes

url            object
score_c         int64
rating_c       object
n_reviews_c     int64
fresh_c         int64
rotten_c        int64
score_u        object
rating_u       object
n_reviews_u    object
director       object
box_office     object
runtime        object
in_theatre     object
on_disc        object
audience       object
studio         object
dtype: object

Several columns are in the wrong dtype. Clean the data!

In [6]:
df.rating_c = df.rating_c.str[:-3]

In [7]:
df.rating_c = pd.to_numeric(df.rating_c)/10

In [8]:
df.rating_u = df.rating_u.str[:-3]
df.rating_u = pd.to_numeric(df.rating_u)/5

In [9]:
df.score_u = pd.to_numeric(df.score_u.str.strip('%'))

In [10]:
df.n_reviews_u = pd.to_numeric(df.n_reviews_u.str.replace(',', ''))

In [11]:
df.box_office = df.box_office.str.replace(',', '').replace('$', '')

In [12]:
df.box_office = pd.to_numeric(df.box_office.str.replace('$', ''))

In [13]:
df.runtime = df.runtime.str.replace('P','')

In [14]:
df.runtime = pd.to_numeric(df.runtime.str.replace('M', ''))

In [15]:
df.in_theatre = pd.to_datetime(df.in_theatre)

In [16]:
df.on_disc = pd.to_datetime(df.on_disc)

In [17]:
df.head()

url  score_c  rating_c  \
0  https://www.rottentomatoes.com/m/resident_evil...       35      0.44   
1   https://www.rottentomatoes.com/m/passengers_2016       31      0.49   
2  https://www.rottentomatoes.com/m/beauty_and_th...       32      0.46   
3  https://www.rottentomatoes.com/m/fantastic_bea...       73      0.68   
4  https://www.rottentomatoes.com/m/absolutely_an...       18      0.37   

   n_reviews_c  fresh_c  rotten_c  score_u  rating_u  n_reviews_u  \
0           85       30        55     51.0      1.18        58527   
1          229       70       159     63.0      0.94        52600   
2           19        6        13     53.0      0.60         3681   
3          271      198        73     79.0      1.40        84702   
4           40        7        33     31.0      0.40         3080   

             director   box_office  runtime in_theatre    on_disc  \
0  Paul W.S. Anderson   26844692.0    106.0 2017-01-27 2017-05-16   
1       Morten Tyldum  100014092.0    116.0 2016-12-21 2017-03-14   
2     Christophe Gans          NaN    112.0 2014-01-03 2017-02-21   
3    David Yates (II)  234018657.0    132.0 2016-11-18 2017-03-28   
4         Terry Jones     673096.0     85.0 2017-05-12 2017-06-27   

                                            audience                 studio  
0           R (for sequences of violence throughout)          Sony Pictures  
1     PG-13 (for sexuality, nudity and action/peril)      Columbia Pictures  
2  PG-13 (for some action violence, peril and fri...                    NaN  
3           PG-13 (for some fantasy action violence)  Warner Bros. Pictures  
4  NR (for language including sexual references, ...     Atlas Distribution

In [18]:
df.dtypes

url                    object
score_c                 int64
rating_c              float64
n_reviews_c             int64
fresh_c                 int64
rotten_c                int64
score_u               float64
rating_u              float64
n_reviews_u             int64
director               object
box_office            float64
runtime               float64
in_theatre     datetime64[ns]
on_disc        datetime64[ns]
audience               object
studio                 object
dtype: object

In [19]:
df['audience_rating'] = df.audience.str.split()

In [20]:
df.audience_rating = [e[0] for e in df.audience_rating]

In [21]:
df.head(2)

url  score_c  rating_c  \
0  https://www.rottentomatoes.com/m/resident_evil...       35      0.44   
1   https://www.rottentomatoes.com/m/passengers_2016       31      0.49   

   n_reviews_c  fresh_c  rotten_c  score_u  rating_u  n_reviews_u  \
0           85       30        55     51.0      1.18        58527   
1          229       70       159     63.0      0.94        52600   

             director   box_office  runtime in_theatre    on_disc  \
0  Paul W.S. Anderson   26844692.0    106.0 2017-01-27 2017-05-16   
1       Morten Tyldum  100014092.0    116.0 2016-12-21 2017-03-14   

                                         audience             studio  \
0        R (for sequences of violence throughout)      Sony Pictures   
1  PG-13 (for sexuality, nudity and action/peril)  Columbia Pictures   

  audience_rating  
0               R  
1           PG-13

In [22]:
df.shape

(1484, 17)

Merge this df with the df obtained with Selenium, containing urls and titles. Merge on url.

In [23]:
df_sel = pd.read_pickle('/Users/aleksandra/ds/metis/project_luther/all_data.pkl')

In [24]:
df_sel.head()

title  \
0            Resident Evil: The Final Chapter   
1                                  Passengers   
2  Beauty And The Beast (La Belle Et La Bête)   
3     Fantastic Beasts And Where To Find Them   
4                         Absolutely Anything   

                                                 url  
0  https://www.rottentomatoes.com/m/resident_evil...  
1   https://www.rottentomatoes.com/m/passengers_2016  
2  https://www.rottentomatoes.com/m/beauty_and_th...  
3  https://www.rottentomatoes.com/m/fantastic_bea...  
4  https://www.rottentomatoes.com/m/absolutely_an...

In [25]:
len(df_sel)

1489

In [26]:
len(df)

1484

In [27]:
df_all_data = df_sel.merge(df, how = 'left', left_on = 'url', right_on='url')

In [28]:
df_all_data.head()

title  \
0            Resident Evil: The Final Chapter   
1                                  Passengers   
2  Beauty And The Beast (La Belle Et La Bête)   
3     Fantastic Beasts And Where To Find Them   
4                         Absolutely Anything   

                                                 url  score_c  rating_c  \
0  https://www.rottentomatoes.com/m/resident_evil...     35.0      0.44   
1   https://www.rottentomatoes.com/m/passengers_2016     31.0      0.49   
2  https://www.rottentomatoes.com/m/beauty_and_th...     32.0      0.46   
3  https://www.rottentomatoes.com/m/fantastic_bea...     73.0      0.68   
4  https://www.rottentomatoes.com/m/absolutely_an...     18.0      0.37   

   n_reviews_c  fresh_c  rotten_c  score_u  rating_u  n_reviews_u  \
0         85.0     30.0      55.0     51.0      1.18      58527.0   
1        229.0     70.0     159.0     63.0      0.94      52600.0   
2         19.0      6.0      13.0     53.0      0.60       3681.0   
3        271.0    198.0      73.0     79.0      1.40      84702.0   
4         40.0      7.0      33.0     31.0      0.40       3080.0   

             director   box_office  runtime in_theatre    on_disc  \
0  Paul W.S. Anderson   26844692.0    106.0 2017-01-27 2017-05-16   
1       Morten Tyldum  100014092.0    116.0 2016-12-21 2017-03-14   
2     Christophe Gans          NaN    112.0 2014-01-03 2017-02-21   
3    David Yates (II)  234018657.0    132.0 2016-11-18 2017-03-28   
4         Terry Jones     673096.0     85.0 2017-05-12 2017-06-27   

                                            audience                 studio  \
0           R (for sequences of violence throughout)          Sony Pictures   
1     PG-13 (for sexuality, nudity and action/peril)      Columbia Pictures   
2  PG-13 (for some action violence, peril and fri...                    NaN   
3           PG-13 (for some fantasy action violence)  Warner Bros. Pictures   
4  NR (for language including sexual references, ...     Atlas Distribution   

  audience_rating  
0               R  
1           PG-13  
2           PG-13  
3           PG-13  
4              NR

In [29]:
len(df_all_data)

1489

In [30]:
len(df_all_data[df_all_data.rating_c.isnull()])

221

221 rows do not correspond in the dfs. I know that 5 movies were not scraped (log Scrapy). What happened with the remaining 226 movies?  

First step: identify the urls that were problematic, then inspect them to see why there is no correspondency.

In [31]:
common = df_sel.merge(df,on='url')
not_in_df = df_sel[(~df_sel.url.isin(common.url))]

In [32]:
not_in_df_sel = df[(~df.url.isin(common.url))]

In [33]:
urls_not_in_df = list(not_in_df.url.sort_values())

In [34]:
urls_not_in_df_sel = list(not_in_df_sel.url.sort_values())

In [35]:
urls_not_in_df

['https://www.rottentomatoes.com/m/1000085-2001_a_space_odyssey',
 'https://www.rottentomatoes.com/m/10004209-tristan_and_isolde',
 'https://www.rottentomatoes.com/m/10004504-ultraviolet',
 'https://www.rottentomatoes.com/m/10004659-arthur',
 'https://www.rottentomatoes.com/m/1000617-aliens',
 'https://www.rottentomatoes.com/m/1000640-all_of_me',
 'https://www.rottentomatoes.com/m/10007142-the_4th_dimension',
 'https://www.rottentomatoes.com/m/1000725-always',
 'https://www.rottentomatoes.com/m/10007898-slipstream',
 'https://www.rottentomatoes.com/m/10007985-happening',
 'https://www.rottentomatoes.com/m/10008502-christmas_carol',
 'https://www.rottentomatoes.com/m/10008588-journey_3d',
 'https://www.rottentomatoes.com/m/10009075-moon',
 'https://www.rottentomatoes.com/m/10009083-land_of_the_lost',
 'https://www.rottentomatoes.com/m/10009151-box',
 'https://www.rottentomatoes.com/m/10009274-priest',
 'https://www.rottentomatoes.com/m/10009460-the_road',
 'https://www.rottentomatoes.co

In [36]:
urls_not_in_df_sel

['https://www.rottentomatoes.com/m/1000085_2001_a_space_odyssey',
 'https://www.rottentomatoes.com/m/10004209_tristan_and_isolde',
 'https://www.rottentomatoes.com/m/10004504_ultraviolet',
 'https://www.rottentomatoes.com/m/10004659_arthur',
 'https://www.rottentomatoes.com/m/1000617_aliens',
 'https://www.rottentomatoes.com/m/1000640_all_of_me',
 'https://www.rottentomatoes.com/m/10007142_the_4th_dimension',
 'https://www.rottentomatoes.com/m/1000725_always',
 'https://www.rottentomatoes.com/m/10007898_slipstream',
 'https://www.rottentomatoes.com/m/10007985_happening',
 'https://www.rottentomatoes.com/m/10008502_christmas_carol',
 'https://www.rottentomatoes.com/m/10008588_journey_3d',
 'https://www.rottentomatoes.com/m/10009075_moon',
 'https://www.rottentomatoes.com/m/10009083_land_of_the_lost',
 'https://www.rottentomatoes.com/m/10009151_box',
 'https://www.rottentomatoes.com/m/10009274_priest',
 'https://www.rottentomatoes.com/m/10009460_the_road',
 'https://www.rottentomatoes.co

In [37]:
print(urls_not_in_df[0])
print(urls_not_in_df_sel[0])

https://www.rottentomatoes.com/m/1000085-2001_a_space_odyssey
https://www.rottentomatoes.com/m/1000085_2001_a_space_odyssey


Some urls have '-' where they have '_' in the other df. We need to correct them...

In [38]:
df['url_uni'] = df.url.str.replace('-', '_')

In [39]:
df_sel['url_uni'] = df_sel.url.str.replace('-', '_')

In [40]:
df_all_data = df_sel.merge(df, on='url_uni')

In [41]:
len(df_all_data)

1484

That's better! Now we can drop the non-unified url columns.

In [42]:
df_all_data.head(2)

title  \
0  Resident Evil: The Final Chapter   
1                        Passengers   

                                               url_x  \
0  https://www.rottentomatoes.com/m/resident_evil...   
1   https://www.rottentomatoes.com/m/passengers_2016   

                                             url_uni  \
0  https://www.rottentomatoes.com/m/resident_evil...   
1   https://www.rottentomatoes.com/m/passengers_2016   

                                               url_y  score_c  rating_c  \
0  https://www.rottentomatoes.com/m/resident_evil...       35      0.44   
1   https://www.rottentomatoes.com/m/passengers_2016       31      0.49   

   n_reviews_c  fresh_c  rotten_c  score_u  rating_u  n_reviews_u  \
0           85       30        55     51.0      1.18        58527   
1          229       70       159     63.0      0.94        52600   

             director   box_office  runtime in_theatre    on_disc  \
0  Paul W.S. Anderson   26844692.0    106.0 2017-01-27 2017-05-16   
1       Morten Tyldum  100014092.0    116.0 2016-12-21 2017-03-14   

                                         audience             studio  \
0        R (for sequences of violence throughout)      Sony Pictures   
1  PG-13 (for sexuality, nudity and action/peril)  Columbia Pictures   

  audience_rating  
0               R  
1           PG-13

In [43]:
df_all_data.drop(['url_x', 'url_y'], axis=1, inplace=True)

In [44]:
df_all_data.head(2)

title  \
0  Resident Evil: The Final Chapter   
1                        Passengers   

                                             url_uni  score_c  rating_c  \
0  https://www.rottentomatoes.com/m/resident_evil...       35      0.44   
1   https://www.rottentomatoes.com/m/passengers_2016       31      0.49   

   n_reviews_c  fresh_c  rotten_c  score_u  rating_u  n_reviews_u  \
0           85       30        55     51.0      1.18        58527   
1          229       70       159     63.0      0.94        52600   

             director   box_office  runtime in_theatre    on_disc  \
0  Paul W.S. Anderson   26844692.0    106.0 2017-01-27 2017-05-16   
1       Morten Tyldum  100014092.0    116.0 2016-12-21 2017-03-14   

                                         audience             studio  \
0        R (for sequences of violence throughout)      Sony Pictures   
1  PG-13 (for sexuality, nudity and action/peril)  Columbia Pictures   

  audience_rating  
0               R  
1           PG-13

Pickle it!